classpath/avian/http/Handler.java

```diff
@@ -10,17 +10,112 @@
 
 package avian.http;
 
-import java.net.URL;
-import java.net.URLStreamHandler;
-import java.net.URLConnection;
+import java.io.BufferedInputStream;
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.ByteArrayInputStream;
 import java.io.IOException;
-import java.io.FileNotFoundException;
-import java.io.File;
-import java.io.FileInputStream;
 import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.io.OutputStreamWriter;
+import java.net.Socket;
+import java.net.URL;
+import java.net.URLConnection;
+import java.net.URLStreamHandler;
+import java.util.HashMap;
+import java.util.Map;
+
+public class Handler extends URLStreamHandler
+{
+    public URLConnection openConnection(URL url) throws IOException
+    {
+        return new HttpURLConnection(url);
+    }
+    
+    class HttpURLConnection extends URLConnection
+    {
+        Socket socket;
+        private BufferedWriter writer;
+        private InputStream bin;
+        private Map<String,String> header = new HashMap<String, String>();
+        private int status;
+        
+        protected HttpURLConnection(URL url)
+        {
+            super(url);
+        }
+
+        @Override
+        public void connect() throws IOException
+        {
+            if(socket == null)
+            {
+                URLConnection con = null;
+                String host = url.getHost();
+                int port =url.getPort();
+                if(port < 0) port = 80;
+                socket = new Socket(host, port);
+                OutputStream out = socket.getOutputStream();
+                writer = new BufferedWriter(new OutputStreamWriter(out));
+                writer.write("GET " + url.getPath() + " HTTP/1.1");
+                writer.write("\nHost: " + host);
+                writer.write("\n\n");
+                writer.flush();
+                bin = new BufferedInputStream(socket.getInputStream());
+                readHeader();
+//                System.out.println("Status: " + status);
+//                System.out.println("Headers: " + header);
+            }
+        }
+
+        private void readHeader() throws IOException
```

Comment:
This implementation looks rather slow and error-prone:

* We probably ought to have a higher limit than 2K on the maximum size of headers (for reference, Apache uses 8K and IIS uses 16K).
* Furthermore, we should probably throw an error if that limit is exceeded, rather than trying to parse the incomplete header block.
* It seems strange to stop the block after any four newline-like characters.  The spec says "\r\n\r\n" should be the terminator, but this would accept "\r\r\r\r", among other oddities.
* Reading headers byte-by-byte off the wire is going to be needlessly slow.  Ideally, this would be implemented by directly parsing from some intermediate buffers, with a state machine to keep track of where we are.  Note that this would require some extra complexity in having to keep track any data we read off the wire but later determined was not part of the header block (i.e. we saw the terminator in the middle of a block; we need to hold on to the rest of it).  We'll need to feed that to the client as the first part of the returned input stream.